torch/csrc/jit/tensorexpr/analysis.h

```diff
@@ -248,6 +248,221 @@ class ModifiesVarChecker : public IRVisitor {
+    for (int i = 0; i < size; i++) {
+      if (trace_.at(i) < compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == trace);
+  }
+
+  std::string getTraceString() {
+    std::string cstr = "";
+    for (int i = 0; i < trace_.size(); i++) {
+      cstr += std::to_string(trace_.at(i));
+      cstr += ":";
+    }
+    return cstr;
+  }
+
+ private:
+  std::vector<int32_t> trace_;
+};
+
+// Traverse the root stmt to identify the position of each stmt in the AST,
+// i.e., a StmtTrace
+class StmtTracer : public IRVisitor {
+ public:
+  StmtTrace getStmtTrace() {
+    return trace_;
+  }
+
+ private:
+  void visit(BlockPtr v) {
+    if (flatten_) {
+      for (StmtPtr s : *v) {
+        s->accept(this);
+      }
+      return;
+    }
+
+    int count = 0;
+    for (StmtPtr s : *v) {
+      trace_.append(count);
+      s->accept(this);
+      trace_.pop();
+      count++;
+    }
+  }
+
+  void visit(CondPtr v) {
+    flatten_ = true;
+    ExprPtr condition = v->condition();
+    StmtPtr true_stmt = v->true_stmt();
+    StmtPtr false_stmt = v->false_stmt();
+    condition->accept(this);
+    if (true_stmt) {
+      true_stmt->accept(this);
+    }
+    if (false_stmt) {
+      false_stmt->accept(this);
+    }
+    flatten_ = false;
+  }
+
+  StmtTrace trace_;
+  bool flatten_ = false;
+};
+
+enum AccMode { READ, WRITE, REDUCE };
+
+// Traverses the IR to identify all reads/writes to a buf, and their positions
+// in the AST which is represented as a StmtTrace
+using BufAccessInfo = std::tuple<StmtPtr, AccMode, StmtTrace>;
+class BufAccesses : public StmtTracer {
+ public:
+  BufAccesses(BufPtr b) : buf_(b) {}
+
+  std::vector<std::tuple<StmtPtr, AccMode, StmtTrace>> accesses() {
+    return accesses_;
+  }
+
+  static std::vector<std::tuple<StmtPtr, AccMode, StmtTrace>> find(
+      StmtPtr s,
+      BufPtr b) {
+    BufAccesses finder(b);
+    s->accept(&finder);
+    return finder.accesses();
+  }
+
+ private:
+  bool readBuffer(StmtPtr s) {
+    auto loads1 = NodeFinder<Load>::find(s);
+    for (auto l : loads1) {
+      if (l->buf() == buf_) {
+        return true;
+      }
+    }
+    auto loads2 = NodeFinder<ExternalCall>::find(s);
+    for (auto l : loads2) {
+      for (auto lb : l->buf_args()) {
+        if (lb == buf_) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool writeBuffer(StmtPtr s) {
+    auto writes1 = NodeFinder<Store>::find(s);
+    for (auto w : writes1) {
+      if (w->buf() == buf_) {
+        return true;
+      }
+    }
+    auto writes2 = NodeFinder<ExternalCall>::find(s);
+    for (auto w : writes2) {
+      if (w->buf() == buf_) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void insertAccesses(StmtPtr s) {
+    bool has_reads = readBuffer(s), has_writes = writeBuffer(s);
+    if (has_reads && has_writes) {
+      auto acc = std::make_tuple(s, AccMode::REDUCE, getStmtTrace());
+      accesses_.push_back(acc);
+      return;
+    }
```

Comment:
Not all Stmts that have reads and writes to a buffer are reduces. If at some point this function would be used for a Block, it could start returning incorrect results. I realize it's not used for Blocks at this moment, but it's preferable to structure the code "defensively" to prevent such situations in future when possible. Would it be possible to refactor this code somehow to avoid that?

Reply:
Oh, I am trying to differentiate different buf accesses: buf read, buf write and buf read+write. `Reduce` operations both read and write the buf so I named the `read+write` mode with `REDUCE`. I can change it to `AccMode::BOTH`.